File: 2891-maximum-beauty-of-an-array-after-applying-operation/2891-maximum-beauty-of-an-array-after-applying-operation.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def maximum_beauty(nums, k)
    nums.sort!
    l = 0
    r = 0
    ans = 0

    while r < nums.size
        while nums[r] - nums[l] > 2 * k
            l += 1
        end
        ans = [ans, r - l + 1].max
        r += 1
    end

    ans
end

File: 2891-maximum-beauty-of-an-array-after-applying-operation/2891-maximum-beauty-of-an-array-after-applying-operation.cpp
class Solution {
public:
    int maximumBeauty(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int l = 0, r = 0, n = nums.size(), ans = 0;

        while(r < n)
        {
            // window expand

            // shrink
            while(nums[r] - nums[l] > 2 * k)
            {
                l++;
            }

            ans = max(ans, r - l + 1);
            r++;
        }

        return ans;
    }
};